semantic-core/src/Data/Core/Pretty.hs

```diff
@@ -31,88 +31,91 @@ showFile = showCore . fileBody
-              pure (lambda <> name x <+> arrow <+> body)
-
-            Frame    -> pure $ primitive "frame"
-            Unit     -> pure $ primitive "unit"
-            Bool b   -> pure $ primitive (if b then "true" else "false")
-            String s -> pure . strlit $ Pretty.viaShow s
-
-            f :$ x -> inParens 11 $ (<+>) <$> go f <*> go x
-
-            If con tru fal -> do
-              con' <- "if"   `appending` go con
-              tru' <- "then" `appending` go tru
-              fal' <- "else" `appending` go fal
-              pure $ Pretty.sep [con', tru', fal']
-
-            Load p   -> "load" `appending` go p
-            Edge Lexical n -> "lexical" `appending` go n
-            Edge Import n -> "import" `appending` go n
-            item :. body   -> inParens 4 $ do
-              f <- go item
-              g <- go body
-              pure (f <> symbol "." <> g)
-
-            lhs := rhs -> inParens 3 $ do
-              f <- go lhs
-              g <- go rhs
-              pure (f <+> symbol "=" <+> g)
+            Rec (Named (Ignored x) b) -> prec 3 . group . nest 2 $ vsep
+              [ keyword "rec" <+> name x
+              , symbol "=" <+> align (withPrec 0 (go (instantiate1 (pure (name x)) b)))
+              ]
+
+            Lam (Named (Ignored x) b) -> prec 3 . group . nest 2 $ vsep
+              [ lambda <> name x, arrow <+> withPrec 0 (go (instantiate1 (pure (name x)) b)) ]
+
+            Record fs -> atom . group . nest 2 $ vsep [ primitive "record", block ", " (map (uncurry keyValue) fs) ]
+
+            Unit     -> atom $ primitive "unit"
+            Bool b   -> atom $ primitive (if b then "true" else "false")
+            String s -> atom . strlit $ viaShow s
+
+            f :$ x -> prec 8 (withPrec 8 (go f) <+> withPrec 9 (go x))
+
+            If con tru fal -> prec 3 . group $ vsep
+              [ keyword "if"   <+> precBody (go con)
+              , keyword "then" <+> precBody (go tru)
+              , keyword "else" <+> precBody (go fal)
+              ]
+
+            Load p -> prec 3 (keyword "load" <+> withPrec 9 (go p))
+            item :. body -> prec 9 (withPrec 9 (go item) <> symbol "." <> name body)
+
+            lhs := rhs -> prec 3 . group . nest 2 $ vsep
+              [ withPrec 4 (go lhs)
+              , symbol "=" <+> align (withPrec 4 (go rhs))
+              ]
 
             -- Annotations are not pretty-printed, as it lowers the signal/noise ratio too profoundly.
             Ann _ c -> go c
-          where bind (Ignored x) f = (,) x <$> go (instantiate1 (pure x) f)
+            statement ->
+              let (bindings, return) = unstatements (Term statement)
+                  statements = toList (bindings :> (Nothing :<- return))
+                  names = zipWith (\ i (n :<- _) -> maybe (pretty @Int i) (name . namedName) n) [0..] statements
+                  statements' = map (prettyStatement names) statements
+              in atom (block "; " statements')
```

Comment:
The `:>>` and `:>>=` cases are both handled by this fallback case, which pretty-prints sequences of semicolon-separated statements in a single `{}`-wrapped block. Unlike the old recursive definition, this has the benefit of not increasing indentation for every statement in right-nested sequences, which greatly improves readability.